File: Server~/src/tools/buildTools.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// ============================================================================
// build_project Tool
// ============================================================================

const buildProjectToolName = 'build_project';
const buildProjectToolDescription = 'Builds the Unity project for a specified target platform. Supports Windows, macOS, Android, iOS, and WebGL.';
const buildProjectParamsSchema = z.object({
  target: z.enum(['StandaloneWindows64', 'StandaloneOSX', 'Android', 'iOS', 'WebGL']).describe('The build target platform'),
  outputPath: z.string().optional().describe('The output path for the build. Defaults to "Builds/<target>/Build<ext>"'),
  developmentBuild: z.boolean().optional().default(false).describe('Whether to create a development build (default: false)'),
  scenes: z.array(z.string()).optional().describe('Array of scene paths to include. Defaults to scenes enabled in Build Settings')
});

/**
 * Registers the build_project tool with the MCP server
 */
export function registerBuildProjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${buildProjectToolName}`);

  server.tool(
    buildProjectToolName,
    buildProjectToolDescription,
    buildProjectParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${buildProjectToolName}`, params);
        const result = await buildProjectHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${buildProjectToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${buildProjectToolName}`, error);
        throw error;
      }
    }
  );
}

async function buildProjectHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.target) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'target' must be provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: buildProjectToolName,
    params: {
      target: params.target,
      outputPath: params.outputPath,
      developmentBuild: params.developmentBuild ?? false,
      scenes: params.scenes
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to build project'
    );
  }

  let text = response.message || 'Build completed';
  text += `\nResult: ${response.buildResult}`;
  text += `\nOutput: ${response.outputPath}`;
  text += `\nDuration: ${response.duration}s`;
  text += `\nErrors: ${response.totalErrors}, Warnings: ${response.totalWarnings}`;

  if (response.errors && response.errors.length > 0) {
    text += '\n\nErrors:\n' + response.errors.join('\n');
  }
  if (response.warnings && response.warnings.length > 0) {
    text += '\n\nWarnings:\n' + response.warnings.join('\n');
  }

  return {
    content: [{
      type: 'text',
      text
    }]
  };
}

// ============================================================================
// get_build_settings Tool
// ============================================================================

const getBuildSettingsToolName = 'get_build_settings';
const getBuildSettingsToolDescription = 'Gets the current build settings including target platform, scenes, and configuration';
const getBuildSettingsParamsSchema = z.object({});

/**
 * Registers the get_build_settings tool with the MCP server
 */
export function registerGetBuildSettingsTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${getBuildSettingsToolName}`);

  server.tool(
    getBuildSettingsToolName,
    getBuildSettingsToolDescription,
    getBuildSettingsParamsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${getBuildSettingsToolName}`, params);
        const result = await getBuildSettingsHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${getBuildSettingsToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${getBuildSettingsToolName}`, error);
        throw error;
      }
    }
  );
}

async function getBuildSettingsHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  const response = await mcpUnity.sendRequest({
    method: getBuildSettingsToolName,
    params: {}
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get build settings'
    );
  }

  let text = `Active Build Target: ${response.activeBuildTarget}\n`;
  text += `Build Target Group: ${response.selectedBuildTargetGroup}\n`;
  text += `Development: ${response.development}\n`;
  text += `Scripting Backend: ${response.scriptingBackend}\n`;
  text += `IL2CPP Configuration: ${response.il2CppCompilerConfiguration}\n\n`;
  text += `Scenes in Build Settings:\n`;

  if (response.scenes && Array.isArray(response.scenes)) {
    for (const scene of response.scenes) {
      text += `  ${scene.enabled ? '[x]' : '[ ]'} ${scene.path}\n`;
    }
  }

  return {
    content: [{
      type: 'text',
      text
    }],
    data: {
      activeBuildTarget: response.activeBuildTarget,
      selectedBuildTargetGroup: response.selectedBuildTargetGroup,
      scenes: response.scenes,
      development: response.development,
      scriptingBackend: response.scriptingBackend,
      il2CppCompilerConfiguration: response.il2CppCompilerConfiguration
    }
  };
}

// ============================================================================
// Combined Registration
// ============================================================================

/**
 * Registers all build tools with the MCP server
 */
export function registerBuildTools(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  registerBuildProjectTool(server, mcpUnity, logger);
  registerGetBuildSettingsTool(server, mcpUnity, logger);
}

File: Editor/Tools/BuildTools.cs
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEngine;
using UnityEditor;
using UnityEditor.Build.Reporting;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for building the Unity project
    /// </summary>
    public class BuildProjectTool : McpToolBase
    {
        public BuildProjectTool()
        {
            Name = "build_project";
            Description = "Builds the Unity project for a specified target platform";
            IsAsync = true;
        }

        public override void ExecuteAsync(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            string targetStr = parameters["target"]?.ToObject<string>();
            string outputPath = parameters["outputPath"]?.ToObject<string>();
            bool developmentBuild = parameters["developmentBuild"]?.ToObject<bool>() ?? false;
            var scenesToken = parameters["scenes"];

            if (string.IsNullOrEmpty(targetStr))
            {
                tcs.TrySetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'target' not provided",
                    "validation_error"
                ));
                return;
            }

            BuildTarget target;
            if (!TryParseBuildTarget(targetStr, out target))
            {
                tcs.TrySetResult(McpUnitySocketHandler.CreateErrorResponse(
                    $"Invalid build target: {targetStr}. Supported: StandaloneWindows64, StandaloneOSX, Android, iOS, WebGL",
                    "validation_error"
                ));
                return;
            }

            // Default output path
            if (string.IsNullOrEmpty(outputPath))
            {
                string ext = GetBuildExtension(target);
                outputPath = $"Builds/{targetStr}/Build{ext}";
            }

            // Ensure output directory exists
            string dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
                Directory.CreateDirectory(dir);

            // Get scenes
            string[] buildScenes;
            if (scenesToken != null && scenesToken.Type == JTokenType.Array)
            {
                buildScenes = scenesToken.ToObject<string[]>();
            }
            else
            {
                buildScenes = EditorBuildSettings.scenes
                    .Where(s => s.enabled)
                    .Select(s => s.path)
                    .ToArray();
            }

            McpLogger.LogInfo($"Building project: target={targetStr}, outputPath={outputPath}, development={developmentBuild}");

            var options = new BuildPlayerOptions
            {
                scenes = buildScenes,
                locationPathName = outputPath,
                target = target,
                options = developmentBuild ? BuildOptions.Development : BuildOptions.None
            };

            var report = BuildPipeline.BuildPlayer(options);

            var errors = new JArray();
            var warnings = new JArray();
            foreach (var step in report.steps)
            {
                foreach (var msg in step.messages)
                {
                    if (msg.type == LogType.Error)
                        errors.Add(msg.content);
                    else if (msg.type == LogType.Warning)
                        warnings.Add(msg.content);
                }
            }

            bool success = report.summary.result == BuildResult.Succeeded;

            tcs.TrySetResult(new JObject
            {
                ["success"] = success,
                ["type"] = "text",
                ["message"] = success
                    ? $"Build succeeded: {outputPath} (duration: {report.summary.totalTime})"
                    : $"Build failed with {report.summary.totalErrors} error(s)",
                ["buildResult"] = report.summary.result.ToString(),
                ["outputPath"] = outputPath,
                ["totalErrors"] = report.summary.totalErrors,
                ["totalWarnings"] = report.summary.totalWarnings,
                ["errors"] = errors,
                ["warnings"] = warnings,
                ["duration"] = report.summary.totalTime.TotalSeconds
            });
        }

        private static bool TryParseBuildTarget(string str, out BuildTarget target)
        {
            switch (str)
            {
                case "StandaloneWindows64": target = BuildTarget.StandaloneWindows64; return true;
                case "StandaloneOSX": target = BuildTarget.StandaloneOSX; return true;
                case "Android": target = BuildTarget.Android; return true;
                case "iOS": target = BuildTarget.iOS; return true;
                case "WebGL": target = BuildTarget.WebGL; return true;
                default: target = default; return false;
            }
        }

        private static string GetBuildExtension(BuildTarget target)
        {
            switch (target)
            {
                case BuildTarget.StandaloneWindows64: return ".exe";
                case BuildTarget.StandaloneOSX: return ".app";
                case BuildTarget.Android: return ".apk";
                default: return "";
            }
        }
    }

    /// <summary>
    /// Tool for getting current build settings information
    /// </summary>
    public class GetBuildSettingsTool : McpToolBase
    {
        public GetBuildSettingsTool()
        {
            Name = "get_build_settings";
            Description = "Gets the current build settings including target platform, scenes, and configuration";
        }

        public override JObject Execute(JObject parameters)
        {
            McpLogger.LogInfo("Executing GetBuildSettingsTool");

            var scenes = EditorBuildSettings.scenes;
            var scenesArray = new JArray();
            foreach (var scene in scenes)
            {
                scenesArray.Add(new JObject
                {
                    ["path"] = scene.path,
                    ["enabled"] = scene.enabled,
                    ["guid"] = scene.guid.ToString()
                });
            }

            var scriptingBackend = PlayerSettings.GetScriptingBackend(
                EditorUserBuildSettings.selectedBuildTargetGroup
            );

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = "Build settings retrieved",
                ["activeBuildTarget"] = EditorUserBuildSettings.activeBuildTarget.ToString(),
                ["selectedBuildTargetGroup"] = EditorUserBuildSettings.selectedBuildTargetGroup.ToString(),
                ["scenes"] = scenesArray,
                ["development"] = EditorUserBuildSettings.development,
                ["scriptingBackend"] = scriptingBackend.ToString(),
                ["il2CppCompilerConfiguration"] = PlayerSettings.GetIl2CppCompilerConfiguration(
                    EditorUserBuildSettings.selectedBuildTargetGroup
                ).ToString()
            };
        }
    }
}
